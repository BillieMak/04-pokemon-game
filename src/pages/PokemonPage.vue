<template>
    <div v-if="!pokemon" class="loading-spiner"></div>
    <div v-else class="container-app">
        <h1>Quien es este Pokemon ?</h1>
        <!-- img -->
        <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
        <!-- Opciones -->
        <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer" />
 
        <template v-if="showAnswer">
            <h2>{{ message }}</h2>
        </template>
    </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'

import getPokemonOptions from '@/helpers/getPokemonOptions';

export default {
    name: 'PokemonPage',
    components: {
        PokemonPicture,
        PokemonOptions,
    },
    data() {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: ''
        }
    },
    methods: {
        newGame() {
            this.message = ''
            this.showPokemon = false
            this.showAnswer = false
            this.pokemon= null
            this.mixPokemonsArray()
        },
        async mixPokemonsArray() {
            this.pokemonArr = await getPokemonOptions()

            const rndInt = Math.floor(Math.random() * this.pokemonArr.length)
            this.pokemon = this.pokemonArr[rndInt]

        },
        checkAnswer(pokemonId) {
            this.showAnswer = true
            if (pokemonId === this.pokemon.id) {
                this.showPokemon = true
                this.showAnswer = true
                setTimeout(() => {
                    this.newGame()
                }, 2000)
                this.message = `Correcto: El pokemon es ${this.pokemon.name}`
            } else {
                setTimeout(() => {
                    this.newGame()
                },700)
                this.message = `Incorrecto: El pokemon es ${this.pokemon.name}`
            }

        }
    },
    mounted() {
        this.mixPokemonsArray()
    }
}
</script>

<style scoped>
.container-app{
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    height: 100vh;
    gap: 20px;
    /* transform: scale(1); */
}
.loading-spiner {
    border: 8px solid #f3f3f3;
    /* Light grey */
    border-top: 8px solid #3498db;
    /* Blue */
    border-radius: 50%;
    width: 50px;
    height: 50px;
    animation: spin 2s linear infinite;
    margin: 0 auto;
    margin-top: 50px;
    position: absolute;
    top: 40%;
    left: 50%;
    transform: translate(-50%, -50%);
}

@keyframes spin {
    0% {
        transform: rotate(0deg);
    }

    100% {
        transform: rotate(360deg);
    }
}
</style>