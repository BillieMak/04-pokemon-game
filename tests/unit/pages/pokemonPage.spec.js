import { shallowMount, mount } from "@vue/test-utils";
import PokemonPage from "@/pages/pokemonPage";
import { pokemonsMock } from "../mocks/pokemons.mock";

describe("PageComponent", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);
  });
  afterEach(() => {
    wrapper.unmount()
  });

  test("Debe de hacer match con el snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("Debe de llamar mixPokemonArray al montar", () => {
    const mixPokemonsArray = jest.spyOn(
      PokemonPage.methods,
      "mixPokemonsArray"
    );
    const wrapper = shallowMount(PokemonPage); 

    expect(mixPokemonsArray).toHaveBeenCalled();
  });

  test("Debe de hacer match con el spanshot cuando cargan los pokemons", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemonsMock,
          pokemon: pokemonsMock[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("Debe de mostrar los componentes de pokemonPicture y pokemonOptions", async() => {
    //todo: debe de existir
    const wrapper =  shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemonsMock,
          pokemon: pokemonsMock[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });

    const pokemonPicture =  wrapper.find("pokemon-picture-stub");
    const pokemonOptions = wrapper.find("pokemon-options-stub");

    expect(pokemonPicture.exists()).toBeTruthy();
    expect(pokemonOptions.exists()).toBeTruthy();

    //* pokemonPicture atrribute pokemonId === 1*/

    const pokemonIdPicture = pokemonPicture.attributes("pokemonid");
    expect(pokemonIdPicture).toBe("1");

    //* pokemonOptions atrribute pokemons toBe true === 1*/

    const pokemons = pokemonOptions.attributes("pokemons");
    expect(pokemons).toBeTruthy();
  });

  test("pruebas con checkAnswer", async() => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemonsMock,
          pokemon: pokemonsMock[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });

    //Propiedades Reactivas de vue
    await wrapper.vm.checkAnswer(1);
    expect(wrapper.vm.showPokemon).toBe(true);

    expect(wrapper.vm.message).toBe(`Correcto: El pokemon es ${pokemonsMock[0].name}`);

    // await wrapper.vm.checkAnswer(100);
    // expect(wrapper.vm.message).toBe(`Incorrecto: El pokemon es ${pokemonsMock[0].name}`);
    // console.log(wrapper.vm.message);

  });

 
});
