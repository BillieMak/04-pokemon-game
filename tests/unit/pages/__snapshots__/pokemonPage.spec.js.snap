// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageComponent Debe de hacer match con el snapshot 1`] = `
<div class="spinner-container">
  <div class="loading-spiner"></div>
</div>
`;

exports[`PageComponent Debe de hacer match con el spanshot cuando cargan los pokemons 1`] = `
<div class="container-app">
  <h1>Quien es este Pokemon ?</h1><!-- img -->
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub><!-- Opciones -->
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
